### PR TITLE
Added unit tests for athena-cloudwatch.

### PR DESCRIPTION
**Pull request by ritiktrianz**
*Issue #, if available:*

*Description of changes:*
 Added unit tests for athena-cloudwatch connector.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java [hunk 1]
```diff
@@ -155,7 +164,7 @@ else if (Integer.valueOf(request.nextToken()) < 3) {
             List<LogGroup> logGroups = new ArrayList<>();
             if (request.nextToken() == null || Integer.valueOf(request.nextToken()) < 3) {
                 for (int i = 0; i < 10; i++) {
-                    LogGroup nextLogGroup = LogGroup.builder().logGroupName("schema-" + String.valueOf(i)).build();
```
Comment (fal-bharadwaj) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java:158]: let's update the name of the test "doListSchemaNames()"and other test methods like "doListTables()", "doGetTable()" etc to follow the naming pattern of other test methods of "do<operation-name>_with<>_returns<>"

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java [hunk 2]
```diff
@@ -414,4 +459,116 @@ public void doGetSplits()
...
+
+            res = handler.doGetQueryPassthroughSchema(allocator, req);
+
+            assertEquals(CATALOG_NAME, res.getCatalogName());
+            assertEquals(tableName, res.getTableName());
+            assertNotNull(res.getSchema());
+            assertEquals(0, res.getSchema().getFields().size());
+
+            verify(mockAwsLogs, times(2)).getQueryResults(nullable(software.amazon.awssdk.services.cloudwatchlogs.model.GetQueryResultsRequest.class));
+    }
```
Comment (fal-bharadwaj) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java:574]: can we add more test cases for edge cases, for example - list tables when log gorups are empty, or when expected output contains empty partitions etc

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchRecordHandlerTest.java [hunk 3]
```diff
@@ -281,6 +269,145 @@ public void doReadRecordsSpill()
         logger.info("doReadRecordsSpill: exit");
     }
 
+    @Test
+    public void readWithConstraint_withPassthrough() throws InterruptedException, TimeoutException {
```
Comment (fal-bharadwaj) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchRecordHandlerTest.java]: lets keep the test names consistent across this connector with other patterns of "do<operation>_with<>_returns/success/failure-result" or something similar.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchRecordHandlerTest.java [hunk 4]
```diff
@@ -281,6 +269,145 @@ public void doReadRecordsSpill()
...
+        assertTrue("Should have at least one remote block", response.getNumberBlocks() > 0);
+        
+        verify(mockAwsLogs, atLeastOnce()).getLogEvents(argThat((GetLogEventsRequest logRequest) ->
+            logRequest.startTime() != null && logRequest.endTime() != null &&
+            logRequest.startTime().equals(1000L) && logRequest.endTime().equals(5000L)
+        ));
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void readRecords_withNullConstraints_throwsException() {
```
Comment (fal-bharadwaj) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchRecordHandlerTest.java]: can we add a test case with empty log events, or where the return is an empty object.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchTableResolverTest.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    @Mock
+    private CloudWatchLogsClient mockAwsLogs;
+
+    @Mock
+    private ThrottlingInvoker mockInvoker;
+
+    private CloudwatchTableResolver resolver;
+
+    @Before
+    public void setUp() throws TimeoutException {
```
Comment (fal-bharadwaj) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchTableResolverTest.java:74]: Can we add tests to test the caching/cache hit logic along with a test for the lambda path casing optimization logic too?

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+            GetQueryResultsResponse response = CloudwatchUtils.getResult(mockInvoker, mockAwsLogs, qptArguments, TEST_LIMIT);
+
+            assertNotNull(response);
+            assertEquals(QueryStatus.COMPLETE, response.status());
+            verify(mockInvoker, times(2)).invoke(any());
+    }
+
+    @Test
+    public void getQueryResults_withNullResponse_returnsNull() {
```
Comment (fal-bharadwaj) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java:165]: let's add test cases for invalid time range, empty log group names, query timeout.

**Review by fal-bharadwaj — CHANGES_REQUESTED** [review]

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java (hunk 1 above)
Comment (ritiktrianz) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java:158]: Hi @fal-bharadwaj , thanks for the review.
 
We originally applied the "do_with<>_returns<>" pattern only to newly added tests. Following your suggestion, we have now updated all existing tests to follow the same pattern.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java (hunk 2 above)
Comment (ritiktrianz) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java:574]: Hi @fal-bharadwaj , added more tests to cover edge cases like list tables when log groups are empty, list schema when results exceeded maximum allowed value, etc.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchRecordHandlerTest.java (hunk 3 above)
Comment (ritiktrianz) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchRecordHandlerTest.java]: Hi @fal-bharadwaj , updated test names to follow MethodName_StateUnderTest_ExpectedBehavior naming convention.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchRecordHandlerTest.java (hunk 4 above)
Comment (ritiktrianz) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchRecordHandlerTest.java]: Hi @fal-bharadwaj  , added more tests to cover these scenarios.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchTableResolverTest.java (hunk 5 above)
Comment (ritiktrianz) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchTableResolverTest.java:74]: Hi @fal-bharadwaj  , added more tests to cover caching hit logic and lambda path casing optimization.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java (hunk 6 above)
Comment (ritiktrianz) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java:165]: Hi @fal-bharadwaj  , added more tests to cover invalid time range, single log group, empty log group names, etc.

**Review by fal-bharadwaj — DISMISSED** [review]
looks good, thanks for making the changes!

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+        // Should handle empty log group names gracefully
+        StartQueryRequest request = CloudwatchUtils.startQueryRequest(emptyLogGroupArgs);
+        assertNotNull(request);
+        assertNotNull(request.logGroupNames());
+        assertEquals(1, request.logGroupNames().size());
+    }
+
+    @Test
+    public void startQueryRequest_withSingleLogGroup_handlesCorrectly() {
```
Comment (christov) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java]: Nitpick: "handles correctly" is ambiguous, which may make it difficult to interpret this test if it fails. Being specific would be more helpful. For example, something like `startQueryRequest_removesLeadingAndTrailingQuotesFromLogGroupNames`

The same comment applies to other test names that use "correctly".

Comment (VenkatasivareddyTR) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java]: Thank you @christov for reviewing the PR. Updated test names as per your suggestion.

**Review by fal-bharadwaj — DISMISSED** [review]

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+        StartQueryResponse response = CloudwatchUtils.getQueryResult(mockAwsLogs, request);
+
+        assertNotNull(response);
+        assertEquals(TEST_QUERY_ID, response.queryId());
+        verify(mockAwsLogs).startQuery(request);
+    }
+
+    @Test
+    public void getQueryResults_withValidRequest_returnsCompleteStatus() {
```
Comment (christov) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java]: I don't think this test tests the functionality of `getQueryResults`. This test checks whether the status is COMPLETE and whether the response is null, but both of these are outside of the functionality of the method. This method has no control over the status or the response--they come from the CW client and `getQueryResults` simply relays back the response object to its caller.

I believe the right way to approach testing this method is to check a) that the method calls the CW client with the correct request object (i.e., an object that contains the query execution ID from the given StartQueryResponse) and b) that it returns the object returned by the CW client.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        assertNotNull(request);
+        assertEquals(TEST_START_TIME, (long) request.startTime());
+        assertEquals(TEST_END_TIME, (long) request.endTime());
+        assertEquals(TEST_QUERY_STRING, request.queryString());
+        String[] expectedLogGroups = new String[]{TEST_LOG_GROUP_1, TEST_LOG_GROUP_2};
+        assertArrayEquals(expectedLogGroups, request.logGroupNames().toArray(new String[0]));
+    }
+
+    @Test
+    public void getQueryResult_withValidRequest_returnsResponseWithQueryId() {
```
Comment (christov) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java]: Similar comment to the [comment](https://github.com/awslabs/aws-athena-query-federation/pull/3042/changes#r2766027937) for testing `getQueryResults` below. I believe this method should test that the CW client is called with the correct request and that the method returns the object the client returned.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+        GetQueryResultsResponse response = CloudwatchUtils.getQueryResults(mockAwsLogs, startQueryResponse);
+
+        assertNotNull(response);
+        assertEquals(QueryStatus.COMPLETE, response.status());
+        verify(mockAwsLogs).getQueryResults(any(GetQueryResultsRequest.class));
+    }
+
+    @Test
+    public void getResult_withValidArguments_returnsCompleteStatusAfterInvocation() throws TimeoutException, InterruptedException {
```
Comment (christov) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java]: Similar comment to the [comment](https://github.com/awslabs/aws-athena-query-federation/pull/3042/changes#r2766027937) for testing `getQueryResults` above. This test is not testing the functionality/responsibility of `getResult`. I think some of the key behaviors of this method include calling the CW client with the correct request (based on the given QPT arguments), returning the correct response, and submitting requests until the status is not COMPLETE.

Ideally, we should also test that the RuntimeException is thrown in the correct scenario, but that part requires making the RESULT_TIMEOUT configurable instead of hard coded, which might be too invasive at this point.

**Review by christov — DISMISSED** [review]
I went over `CloudwatchUtilsTest` (haven't gotten over the remaining classes yet) and left some comments. One general observation is that some of the tests miss to test the functionality/behavior of the method under test and/or test functionality that belongs to the collaborators of the method under test, not to the method itself. Examples are included in the comments on `CloudwatchUtilsTest`. Besides addressing the comments on that class, I recommend reviewing the tests in other classes from that perspective.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java (hunk 8 above)
Comment (ritiktrianz) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java]: Hi @christov , thank you for the review; we’ve updated the test to ensure it properly validates the responsibilities of getQueryResults by verifying that the method returns exactly the response provided by the CloudWatch client and that the client is invoked with a request containing the correct query execution ID from the StartQueryResponse, keeping the test focused on the method’s core functionality.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java (hunk 9 above)
Comment (ritiktrianz) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java]: Hi @christov , the test was already verifying that the CloudWatch client was called with the exact request object (verify(mockAwsLogs).startQuery(request)), and I’ve now updated it further to verify that the method returns the exact response from the CloudWatch client.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java (hunk 10 above)
Comment (ritiktrianz) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchUtilsTest.java]: Updated the test to verify the polling behavior, calling the CW client with the correct request and to verify that correct response is returned.

**Comment by ritiktrianz** [comment]
> I went over `CloudwatchUtilsTest` (haven't gotten over the remaining classes yet) and left some comments. One general observation is that some of the tests miss to test the functionality/behavior of the method under test and/or test functionality that belongs to the collaborators of the method under test, not to the method itself. Examples are included in the comments on `CloudwatchUtilsTest`. Besides addressing the comments on that class, I recommend reviewing the tests in other classes from that perspective.

Hi @christov , Thanks for the review, I’ve addressed the comments on CloudwatchUtilsTest, updated the tests to ensure they focus on the functionality of the methods under test, and also verified and made updates in the other classes tests wherever required.

**Review by burhan94 — COMMENTED** [review]
## Code Review

### ✅ Looks Good

Excellent query passthrough coverage and good edge cases. Proper ArgumentCaptor usage.

### 🟢 Minor
- `Mockito.lenient()` usage can hide test issues — use sparingly.
- Magic number 50_000 should be a named constant.

**Comment by ritiktrianz** [comment]
> ## Code Review
> ### ✅ Looks Good
> 
> Excellent query passthrough coverage and good edge cases. Proper ArgumentCaptor usage.
> ### 🟢 Minor
> 
>     * `Mockito.lenient()` usage can hide test issues — use sparingly.
> 
>     * Magic number 50_000 should be a named constant.

Hi @burhan94  — thanks for the review.
 
• Mockito.lenient(): Its usage here does not hide any real test issues. It is applied only to avoid Mockito strict‑stubbing warnings for stubs that may not be used in every test (e.g., shared setup or unused mock overloads).
• Constants: Magic values (such as `50_000`) have now been replaced with named constants in the tests for better readability and maintainability.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java [hunk 11]
```diff
@@ -234,15 +244,52 @@ else if (Integer.valueOf(request.nextToken()) < 3) {
...
+        verify(mockAwsLogs, times(1)).describeLogStreams(argThat((DescribeLogStreamsRequest request) ->
+            request.nextToken() != null && request.nextToken().equals("test-token") &&
+            request.limit() != null && request.limit().equals(50)
+        ));
+    }
+
+    @Test
+    public void doGetTable_withValidTableName_returnsTableSchemaWithColumns()
     {
         logger.info("doGetTable - enter");
```
Comment (christov) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java]: The method name has changes from `doGetTable` to `doGetTable_withValidTableName_returnsTableSchemaWithColumns`, but the log messages have not been updated accordingly. Could you please make these updates and also review the rest of the PR for occurrences of the same issue?

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java [hunk 12]
```diff
@@ -234,15 +244,52 @@ else if (Integer.valueOf(request.nextToken()) < 3) {
...
+        assertTrue("Should contain paginated table", res.getTables().contains(new TableName("schema-1", "paginated-table-0")));
+
+        verify(mockAwsLogs, times(1)).describeLogStreams(argThat((DescribeLogStreamsRequest request) ->
+            request.nextToken() != null && request.nextToken().equals("test-token") &&
+            request.limit() != null && request.limit().equals(50)
+        ));
+    }
+
+    @Test
+    public void doGetTable_withValidTableName_returnsTableSchemaWithColumns()
```
Comment (christov) [inline athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java]: I don't think the name of the test reflects what the test is actually validating. The assertions are not checking whether the method under test returns a schema with columns.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java (hunk 11 above)
Comment (ritiktrianz) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java]: Hi @christov , thanks for the review.
We have now updated the corresponding log messages to match the test names across the connector.

File: athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java (hunk 12 above)
Comment (ritiktrianz) [reply athena-cloudwatch/src/test/java/com/amazonaws/athena/connectors/cloudwatch/CloudwatchMetadataHandlerTest.java]: Hi @christov , updated the test name to reflect the actual assertions being made.

**Review by hareeshkumar-t — APPROVED** [review]

**Review by AbdulR3hman — APPROVED** [review]
